# Homework 4

* Assigned: 4/8 Fri
* Due: 4/22 Fri 10 AM
* Value: 3.75% of your grade
* **Remember: homeworks are to be completed individually**

In this part of the problem set, you will examine query plans that PostgreSQL uses to execute queries, and try to understand
why it produces the plan it does for a certain query. The data set you will use has the same schema as the `iowa` dataset in HW3.

**NOTE: The iowa table is fairly large with lots of rows, so please try not to run too many generic queries like “SELECT * FROM iowa”. They take a long time to execute, and slow down the database for everyone else. Please see Jupyter notification for shutting down queries.**   

**EXPLAINs are fine since they don't actually execute the queries. When running a query, always use LIMIT clauses and/or selection filters to reduce the number of rows produced.**

### Jupyter Notes: _Read these carefully_

* You **may** create new IPython notebook cells to use for e.g. testing, debugging, exploring, etc.- this is encouraged in fact!- **just make sure that you run the final cell to submit your results**
  * you can press shift+enter to execute to code in the cell that your cursor is in.
* When you see `In [*]:` to the left of the cell you are executing, this means that the code / query is _running_. Please wait for the execution to complete
    * **If the cell is hanging- i.e. running for too long: you can restart the kernel**
    * To restart kernel using the menu bar: "Kernel >> Restart >> Clear all outputs & restart"), then re-execute cells from the top
* _Have fun!_

### Before Starting
**Please run the following cells to allow COMPLETE output for EXPLAIN query, and connect to db**

In [1]:
!pip3 install sqlalchemy # ORM for databases
!pip3 install ipython-sql # SQL magic function

In [2]:
%load_ext sql

In [3]:
%sql postgresql://student:w4111student@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111

/usr/local/lib/python3.7/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


'Connected: student@w4111'

In Part II, we have provided you with the following indexes:

    Indexes:
      "iowa_cat_btree" btree (category)
      "iowa_date" btree (date)
      "iowa_dt_store_item_vendor_tree" btree (date, store, item, vendor)
      "iowa_store_hash" hash (store)
      "iowa_store_item_vendor_dt_tree" btree (store, item, vendor, date)
      "iowa_store_tree" btree (store)
      "iowa_vendor_hash" hash (vendor)
      "iowa_vendor_tree" btree (vendor)
      "iowa_zip_hash" hash (zipcode)
      "iowa_zip_tree" btree (zipcode)

You can view the indexes of iowa using the following commands:

In [4]:
%%sql
select *
from pg_indexes
where schemaname='public' and tablename='iowa';

 * postgresql://student:***@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111
10 rows affected.


[('public', 'iowa', 'iowa_zip_tree', None, 'CREATE INDEX iowa_zip_tree ON public.iowa USING btree (zipcode)'),
 ('public', 'iowa', 'iowa_zip_hash', None, 'CREATE INDEX iowa_zip_hash ON public.iowa USING hash (zipcode)'),
 ('public', 'iowa', 'iowa_vendor_tree', None, 'CREATE INDEX iowa_vendor_tree ON public.iowa USING btree (vendor_no)'),
 ('public', 'iowa', 'iowa_vendor_hash', None, 'CREATE INDEX iowa_vendor_hash ON public.iowa USING hash (vendor_no)'),
 ('public', 'iowa', 'iowa_store_item_vendor_dt_tree', None, 'CREATE INDEX iowa_store_item_vendor_dt_tree ON public.iowa USING btree (store, itemno, vendor_no, date)'),
 ('public', 'iowa', 'iowa_dt_store_item_vendor', None, 'CREATE INDEX iowa_dt_store_item_vendor ON public.iowa USING btree (date, store, itemno, vendor_no)'),
 ('public', 'iowa', 'iowa_store_tree', None, 'CREATE INDEX iowa_store_tree ON public.iowa USING btree (store)'),
 ('public', 'iowa', 'iowa_store_hash', None, 'CREATE INDEX iowa_store_hash ON public.iowa USING hash (store)'),
 ('public', 'iowa', 'iowa_date', None, 'CREATE INDEX iowa_date ON public.iowa USING btree (date)'),
 ('public', 'iowa', 'iowa_cat_btree', None, 'CREATE INDEX iowa_cat_btree ON public.iowa USING btree (category)')]

### A Quick Example

To understand what query plan is being used, PostgreSQL includes the `EXPLAIN` command. 

It prints the plan for a query, including all of the physical operators and access methods being used. 
For example, the following SQL command displays the query plan for the SELECT:

In [5]:
%%sql 
EXPLAIN SELECT * FROM iowa WHERE vendor_no = 0;

 * postgresql://student:***@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111
4 rows affected.


[('Bitmap Heap Scan on iowa  (cost=19.59..3535.05 rows=925 width=1534)',),
 ('  Recheck Cond: (vendor_no = 0)',),
 ('  ->  Bitmap Index Scan on iowa_vendor_tree  (cost=0.00..19.36 rows=925 width=0)',),
 ('        Index Cond: (vendor_no = 0)',)]

For example, this is a query plan with no branches. It first runs a Bitmap Index Scan using the index iowa_vendor_tree, which is a Btree index, and the condition vendor_no = 0.  It _estimates_ that there would be 925 rows that match the condition.   

The results are then fed into a Bitmap Heap Scan, which gathers all the tuple ids from the index scan together, sorts the tuple ids by the pages the tuples are stored in, and reads the data pages as a single scan while rechecking the vendor condition.

Don't worry about the heap scan too much. We mainly care that the query uses the iowa_vendor_tree index. You should also keep in mind that leaves of the BTree index do not store actual tuples (i.e. it is a secondary index, not a primary index).

For more details of how to interpret the result, please check https://www.postgresql.org/docs/current/using-explain.html.

**HINT: In some questions it is necessary to provide with some selectivity of information, so you may want to use COUNT function to write some queries from time to time.**

In [6]:
%%sql
SELECT COUNT(*) FROM iowa;

 * postgresql://student:***@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111
1 rows affected.


[(1000000,)]

### Part II

**Q1**: Run `EXPLAIN` on the following query and explain in your own words (in a few sentences) the query plan that PostgreSQL picked (we are expecting something similar to the given example above).

In [7]:
%%sql
EXPLAIN SELECT * FROM iowa WHERE zipcode = '10027';

 * postgresql://student:***@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111
4 rows affected.


[('Bitmap Heap Scan on iowa  (cost=18.56..3043.23 rows=792 width=1534)',),
 ("  Recheck Cond: (zipcode = '10027'::text)",),
 ('  ->  Bitmap Index Scan on iowa_zip_tree  (cost=0.00..18.36 rows=792 width=0)',),
 ("        Index Cond: (zipcode = '10027'::text)",)]

In [8]:
## please answer between the quotes
a1="""

"""

**Q2**: What did PostgreSQL estimate the number of resulting rows to be and what is the actual number of rows?  
   
Why is there a difference?
_Hint_: Think about how optimizer performs evaluation.

In [9]:
%%sql
-- run this query to get actual number returned
SELECT COUNT(*) FROM iowa WHERE zipcode = '10027';

 * postgresql://student:***@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111
1 rows affected.


[(0,)]

In [10]:
## please answer between the quotes
a2="""

"""

**Q3**: Run `EXPLAIN` on the slightly different query below.  What index does the query use and why is
   it the same or different than the result of Q1?


In [11]:
%%sql
EXPLAIN SELECT * FROM iowa WHERE zipcode = '10027' LIMIT 1;

 * postgresql://student:***@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111
3 rows affected.


[('Limit  (cost=0.00..4.04 rows=1 width=1534)',),
 ('  ->  Index Scan using iowa_zip_hash on iowa  (cost=0.00..3197.86 rows=792 width=1534)',),
 ("        Index Cond: (zipcode = '10027'::text)",)]

In [12]:
## please answer between the quotes
a3="""

"""

**Q4**: Run `EXPLAIN` on the following slightly different queries.  Why does the database choose those plans? What are the main reasons for different plans?


In [13]:
%%sql 
-- Q4A
EXPLAIN SELECT * FROM iowa WHERE '50056' < zipcode AND zipcode < '50058';

 * postgresql://student:***@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111
2 rows affected.


[('Index Scan using iowa_zip_tree on iowa  (cost=0.42..8.45 rows=1 width=1534)',),
 ("  Index Cond: (('50056'::text < zipcode) AND (zipcode < '50058'::text))",)]

In [14]:
%%sql
-- Q4B
EXPLAIN SELECT * FROM iowa WHERE '50056' < zipcode AND zipcode < '52726';

 * postgresql://student:***@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111
2 rows affected.


[('Seq Scan on iowa  (cost=0.00..215000.00 rows=850022 width=1534)',),
 ("  Filter: (('50056'::text < zipcode) AND (zipcode < '52726'::text))",)]

In [15]:
## please answer between the quotes
a4="""

"""

**Q5**: Try the following two EXPLAIN queries (Q5A, Q5B). Why do they have equivalent query plans despite the fact that Q5B has an equality condition?
_Hint_: Think from selectivity and cost statistics yield by `EXPLAIN` query in Q4 and Q5.

In [16]:
%%sql
--Q5A
EXPLAIN SELECT * FROM iowa WHERE 4500 < store AND store < 8000;

 * postgresql://student:***@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111
4 rows affected.


[('Bitmap Heap Scan on iowa  (cost=3092.04..198273.22 rows=145523 width=1534)',),
 ('  Recheck Cond: ((4500 < store) AND (store < 8000))',),
 ('  ->  Bitmap Index Scan on iowa_store_tree  (cost=0.00..3055.66 rows=145523 width=0)',),
 ('        Index Cond: ((4500 < store) AND (store < 8000))',)]

In [17]:
%%sql 
--Q5B
EXPLAIN SELECT * FROM iowa WHERE store = 2633;

 * postgresql://student:***@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111
4 rows affected.


[('Bitmap Heap Scan on iowa  (cost=176.24..30741.06 rows=9267 width=1534)',),
 ('  Recheck Cond: (store = 2633)',),
 ('  ->  Bitmap Index Scan on iowa_store_tree  (cost=0.00..173.93 rows=9267 width=0)',),
 ('        Index Cond: (store = 2633)',)]

In [18]:
## please answer between the quotes
a5="""

"""

**Q6**: Consider if we inserted a large batch of new records into the table.  What is the difference in the amount of time it takes change if the table did not contain any indexes, and if the table did contain the indexes?

In [19]:
## please answer between the quotes...
a6="""

"""

## Part II Submission

To submit your answers, please go to Gradescope -> 2020S W4111 -> HW4. Copy and paste your answers into the part2 submissions.

In [20]:
result = {
    "a1":a1,
    "a2":a2,
    "a3":a3,
    "a4":a4,
    "a5":a5,
    "a6":a6,
}

for key in result:
  print(key + ":", result[key])

a1: 


a2: 


a3: 


a4: 


a5: 


a6: 


